File: leetcode/cpp/arrays_and_hashing/lc_36_valid_sudoku.cpp
#include <vector>
#include <unordered_set>
#include <unordered_map>

/**
 * https://leetcode.com/problems/valid-sudoku/description/
 *
 * Time complexity: O(1) because sudoku is fixed size board
 * Space complexity: O(1) because sudoku is fixed size board
 */
class Solution {
public:
    bool isValidSudoku(std::vector<std::vector<char>>& board) {
        // setup our rowMap to keep track of which numbers exist in rows
        // key = row index, value = set of numbers in that row
        std::unordered_map<int, std::unordered_set<int>> rowMap;
        // setup our colMap to keep track of which numbers exist in cols
        // key = col index, value = set of numbers in that col
        std::unordered_map<int, std::unordered_set<int>> colMap;
        // setup our boxMap to keep track of which numbers exist in each 3x3 box
        // key = flattened (box row, box col) index, value = set of numbers in that 3x3 box
        std::unordered_map<int, std::unordered_set<int>> boxMap;

        // get the row and col lengths
        const int ROW_LEN = board.size();
        const int COL_LEN = board[0].size();

        // loop through all rows
        for (int row{0}; row < ROW_LEN; ++row) {
            // loop through all cols
            for (int col{0}; col < COL_LEN; ++col) {
                // if the board is not occupied by a number, skip it
                if (board[row][col] == '.') {
                    continue;
                }

                // if the rowMap set contains this number
                // we found a duplicate so return false
                if (rowMap[row].contains(board[row][col])) {
                    return false;
                }

                // if the colMap set contains this number
                // we found a duplicate so return false
                if (colMap[col].contains(board[row][col])) {
                    return false;
                }

                // compute the 3x3 box row and box col
                const int boxRow = row / 3;
                const int boxCol = col / 3;

                // compute row major indexing for the box row and box col
                // formula is index = i * total_num_columns + j
                // so we can use an int as our key for boxMap
                const int boxMapKey = boxRow * 3 + boxCol;

                // if the boxMap contains this number
                // we found a duplicate so return false
                if (boxMap[boxMapKey].contains(board[row][col])) {
                    return false;
                }

                // finally, insert the number into each map on each iteration
                rowMap[row].insert(board[row][col]);
                colMap[col].insert(board[row][col]);
                boxMap[boxMapKey].insert(board[row][col]);
            }
        }

        // after checking all numbers and finding no duplicates
        // we can return true
        return true;
    }
};

File: leetcode/arrays_and_hashing/lc_36_valid_sudoku.py
from collections import defaultdict
from typing import List


'''
https://leetcode.com/problems/valid-sudoku/description/
'''
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # create ditionary mapping row index -> set of numbers in row
        row_dictionary = defaultdict(set)
        # create dictionary mapping col index -> set of numbers in col
        col_dictionary = defaultdict(set)
        # create dictionary mapping 
        # (sub_box_row, sub_box_col) -> set of numbers in sub_box
        # note that 0 <= sub_box_row, sub_box_col < 2
        sub_box_dictionary = defaultdict(set)

        # store the row and col length of the grid
        # this is always going to be 9x9 for a sudoku grid
        row_len = 9
        col_len = 9

        # loop through all rows in the sudoku board
        for row in range(row_len):
            # loop through all cols in the sudoku board
            for col in range(col_len):
                # get the current cell
                current_cell = board[row][col]

                # we only want to proceed if there is a number
                # in the current cell, it can't be '.'
                if current_cell != '.':
                    # check if the current cell exists in our row already
                    if current_cell in row_dictionary[row]:
                        # if it does, return False
                        return False
                    # else it doesn't exist
                    else:
                        # so add it our row dictionary
                        row_dictionary[row].add(current_cell)

                    # check if the current cell exists in our col dictionary
                    if current_cell in col_dictionary[col]:
                        # if it does, return False
                        return False
                    # else it doesn't exist
                    else:
                        # so add it to our col dictionary
                        col_dictionary[col].add(current_cell)

                    # map the current row to the appropriate sub_box_row
                    # between [0, 3)
                    sub_box_row = row // 3
                    # map the current col to the appropriate sub_box_col
                    # between [0, 3)
                    sub_box_col = col // 3

                    # check if the current cell exists in our sub_box already
                    if current_cell in sub_box_dictionary[(sub_box_row, sub_box_col)]:
                        # if it does, return False
                        return False
                    # else it doesn't exist
                    else:
                        # so add it our sub_box_dictionary
                        sub_box_dictionary[(sub_box_row, sub_box_col)].add(current_cell)

        # if we loop through the entire grid without
        # detecting any duplicates, the grid is valid
        # so return true
        return True
